src/Storages/StorageURL.cpp

```diff
@@ -185,11 +187,14 @@ namespace
                         std::string user_info = request_uri.getUserInfo();
                         if (!user_info.empty())
                         {
-                            std::size_t n = user_info.find(':');
+                            char decoded_user_info[user_info.length() + 1];
+                            size_t decode_len = decodeURL(user_info.c_str(), user_info.length(), decoded_user_info);
+                            std::string url_decoded_user_info(decoded_user_info, decode_len);
+                            std::size_t n = url_decoded_user_info.find(':');
```

Comment:
I think we should first split then decode usernama & password separately. Username (or password) can have encoded colon. If the colon is missing - those user_info should be considered as a username, i.e.
```cpp
    std::string::size_type pos = user_info.find(':');
	if (pos != std::string::npos)
	{
		std::string decoded_user;
		std::string decoded_password;
		URI::decode(user_info.substr(0, n), decoded_user);
		URI::decode(user_info.substr(n + 1), decoded_password);
        credentials.setUsername(decoded_user);
        credentials.setPassword(decoded_password);
	}
    else {
		std::string decoded_user;
		URI::decode(user_info, decoded_user);
        credentials.setUsername(decoded_user);   
    }
```

Reply:
I'm afraid that may not work. Here's an example, a username is user23 and a password is 34sw/2.
The url encode result(username:password) is user23%3A34sw%2F2. You even can not find a colon symbol.

Reply:
That example is incorrect. Try it if curl / wget / real browser to see it. 

It should be `http://user23:34sw%2F2@host`

```
curl -D- -vvv 'http://default%3A123@localhost:8123/?query=SELECT%201'
*   Trying 127.0.0.1:8123...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8123 (#0)
* Server auth using Basic with user 'default:123'

curl -D- -vvv 'http://default:123@localhost:8123/?query=SELECT%201'
*   Trying 127.0.0.1:8123...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8123 (#0)
* Server auth using Basic with user 'default'
```